ouroboros-network/src/Ouroboros/Network/Mux/Interface.hs

```diff
@@ -56,81 +55,139 @@ import           Ouroboros.Network.Mux.Types
 --   a function that runs the mux layer on it.
 --
 
+data AppType where
+    ClientApp          :: AppType
+    ServerApp          :: AppType
+    ClientAndServerApp :: AppType
+
+
+type family HasClient (appType :: AppType) :: Bool where
+    HasClient ClientApp          = True
+    HasClient ServerApp          = False
+    HasClient ClientAndServerApp = True
+
+type family HasServer (appType :: AppType) :: Bool where
+    HasServer ClientApp          = False
+    HasServer ServerApp          = True
+    HasServer ClientAndServerApp = True
+
 -- |
--- Specification for peers of a protocol.  This type instructs the multiplexing
--- layer to run a given client \/ server peers.
+-- Application run by mux layer.
+--
+-- * enumeration of client application, e.g. a wallet application communicating
+--   with a node using ChainSync and TxSubmission protocols; this only requires
+--   to run client side of each protocol.
 --
-data MuxPeer m where
-  -- |
-  -- A non pipeliend peer together with a codec.
-  --
-  OnlyClient
-    :: forall ps (st :: ps) m a.
-       Tracer m (TraceSendRecv ps)
-    -> Codec ps CBOR.DeserialiseFailure m ByteString
-    -> Peer ps AsClient st m a
-    -> MuxPeer m
-
-  -- |
-  -- A pipelined peer together with a codec.
-  --
-  OnlyPipelinedClient
-    :: forall ps (st :: ps) m a.
-       Natural
-    -> Tracer m (TraceSendRecv ps)
-    -> Codec ps CBOR.DeserialiseFailure m ByteString
-    -> PeerPipelined ps AsClient st m a
-    -> MuxPeer m
-
-  -- |
-  -- Server peer with a codec
-  OnlyServer
-    :: forall ps (st :: ps) m a.
-       Tracer m (TraceSendRecv ps)
-    -> Codec ps CBOR.DeserialiseFailure m ByteString
-    -> Peer ps AsServer st m a
-    -> MuxPeer m
-
-  -- |
-  -- Client and server peers with the corresponding codec.
-  --
-  ClientAndServer
-    :: forall ps (st :: ps) m a.
-       Tracer m (TraceSendRecv ps)
-    -> Codec ps CBOR.DeserialiseFailure m ByteString
-    -> Peer ps AsClient st m a
-    -> Peer ps AsServer st m a
-    -> MuxPeer m
-
-  -- |
-  -- Pipelined client and a server with the correspnding codec.
-  --
-  PipelinedClientAndServer
-    :: forall ps (st :: ps) m a.
-       Natural
-    -> Tracer m (TraceSendRecv ps)
-    -> Codec ps CBOR.DeserialiseFailure m ByteString
-    -> PeerPipelined ps AsClient st m a
-    -> Peer ps AsServer st m a
-    -> MuxPeer m
+-- * enumeration of server applications: this application type is mostly useful
+--   tests.
+--
+-- * enumeration of both client and server applications, e.g. a full node
+--   serving downstream peers using server side of each protocol and getting
+--   updates from upstream peers using client side of each of the protocols.
+--
+data MuxApplication (appType :: AppType) ptcl m  where
```

Comment:
What if we just had a set of peer/peer/codec pairs, where the 2 peers are complementary? Something like this:

```
data MiniProtocol ps st failure m = MiniProtocol
  { mpCodec :: Codec ps failure m ByteString
  , mpAsClient :: Peer ps AsClient st m a
  , mpAsServer :: Peer ps AsServer st m a
  }

-- A MiniProtocol with existential protocol and initial state.
data SomeMiniProtocol failure m where
  SomeMiniProtocol :: MiniProtocol ps st failure m -> SomeMiniProtocol failure m

-- 0 or more MiniProtocols, keyed on their ptcl enumeration
data MuxApplication ptcl failure m where
  MuxApplication :: Map ptcl (SomeMiniProtocol failure m) -> MuxApplication failure m
```

This is a tighter description than `ptcl -> Channel m ByteString -> m ()`. From a `MuxApplication`, you can choose whether to run just the client sides, just the server sides, or both. Those `Bool`-valued type families `HasServer` and `HasClient` are no longer necessary: if you have the server, you also have the client. And of course it ought to be that way! We never define a client without also defining the server. There's no sense in not including both here for a unified description of the entire mini-protocol.